### ISSUE TITLE
Quartz scheduler randomly stops firing tasks

### ISSUE THREAD

**Issue by keshimp** [quartznet/quartznet#644]
Current version : 3.0.6
Current .NET framework : .NET Framework 4.6.2

Issue: Scheduler stops at random intervals and discontinues scheduler - Must restart app pool to re-initiate / build task triggers to continue running jobs.

Our system currently uses around 30 background tasks at different times, from every minute to once a month.  We have tried using the simple schedule, DailyTimeInterval schedule, and crons.

Here are some examples of the code we use in each iteration:

This wrapper function builds all of our tasks by type and trigger:
		private static async Task ScheduleAsync(BackgroundTaskType type, ITrigger trigger)
		{
			var jobKey = JobKey.Create(type.ToString());
			var job = JobBuilder.Create<QuartzJob>().WithIdentity(jobKey).Build();
			ReschedulableJobs.Add(type, job);
			scheduler.ScheduleJob(job, trigger);
		}
		
Simple: 
		await Schedule(BackgroundTaskType.HelloWorld,
			TriggerBuilder.Create()
			.WithSimpleSchedule(x => x.WithIntervalInMinutes(6).RepeatForever()).StartNow().Build()
			);
		
DailyTimeInterval: 
		await ScheduleAsync(BackgroundTaskType.TheTask,
			TriggerBuilder.Create()
			.WithDailyTimeIntervalSchedule(
				x => x.WithIntervalInMinutes(60)
					.OnEveryDay()
					.StartingDailyAt(TimeOfDay.HourAndMinuteOfDay(0, 0)) // midnight central
					.InTimeZone(TimeZoneInfo.Local)
			).Build());
		
Cron:
		await ScheduleAsync(BackgroundTaskType.TheTask,
			TriggerBuilder.Create()
				.WithCronSchedule("0 0 12 * * ?")
			.Build()); // fire every day at noon
				

				
await scheduler.Start();

Currently, the simple schedule will fire when we want it to without issue, but on deploy it will fire multiple of each task simultaneously - and because the schedule is based on time since deploy, the actual run times can be
erratic and unpredictables - not what we want.  So that brought us to the daily time interval schedule, which gave us exactly what we wanted, however, the scheduler would randomly stop firing tasks - assuming this is due to some
sort of deadlock somewhere.  We don't get any exceptions, any warnings, nothing.  All that happens is, tasks will cease to fire, and we will need to restart the app pool.  This is a huge issue for us, and the only workaround
that we have found is by writing an application to restart the app pool after 15 minutes of no tasks running.  Obviously not ideal.

Crons will not fire at all with our current implementation.  That was our last hope at getting some consistency.

Right now the issue is that we would like to deviate from simple scheduling BUT ALSO would like our scheduler to be consistent and run 24/7 without any hiccups.  The hangup with tasks ceasing to fire is a huge deterrent to this.

Extra question:
Is this due to all tasks being async? (We used to run with mostly simple scheduler w/o async code.)

Thanks in advance,

**Comment by j03wang** [comment]
I'm currently experiencing the same issue with 3.0.5--I have a few jobs that are scheduled to fire every minute--once every few days, I'll notice that one of the jobs will stop firing, requiring me to restart the process.

**Comment by cvocvo** [comment]
Also experiencing the same issue on 3.0.6

**Comment by lahma** [comment]
The same reply I gave on mailing list, some crucial information is missing about hosting and setup:

* What version of IIS, is always on mode used? 
* Where do you start your scheduler - Application_Start?
* Do you have handle for scheduler kept as global static variable? 

https://www.quartz-scheduler.net/documentation/faq.html#quartz-in-web-environment

**Comment by j03wang** [comment]
My instance of Quartz is in a standalone .Net Core console app.

**Comment by lahma** [comment]
If someone can take a process dump from the hang/unresponsiveness that would be super, as this is all guesswork otherwise. Unless there's an easy repro app.

**Comment by keshimp** [comment]
Thank you for the reply - 

The scheduler is indeed started in Application_Start
`JobScheduler.StartAsync();`

IIS verion is - IIS Version 10.0.14393.0 on Server 2016 (latest patches applied) and AppPool Start Mode is set to "OnDemand" instead of "AlwaysRunning".

No I do not have the scheduler handler as a global static variable.

**Comment by lahma** [comment]
@keshimp please try with scheduler as static variable (maybe global.cs) so it won't go out of scope and then if the pool recycles or shuts down it's normal that Quartz is no longer running until the first request.

**Comment by keshimp** [comment]
Thanks for the prompt reply - I've added your suggestions and am waiting to see the results, will update when I know more.

**Comment by keshimp** [comment]
Quick report back - I made a change when we register the scheduler - all of my scheduler code was async, but I wasn't awaiting the actually scheduler code to finish (which visual studio gave me a nice warning about) so I assumed there was a possible race condition... and the last couple of days, no tasks had stopped so I assumed that was the fix.  However, I just randomly stopped again this morning

I haven't tried setting the IIS to "always on" - and I have to admit I'm a bit confused when you say "until first request" - do you mean web request?  As in a user making an http request?  I have a task that runs every minute, so I assume that should be running regardless of whether or not a user has made a request  within the last minute or even if it wasn't for a few minutes.  If that's not the case, then would setting IIS to "always on" fix that issue?

Would any other changes need to be made to my scheduling system if I make a global static variable for the scheduler in global.cs?  

Appreciate the help greatly,

**Comment by lahma** [comment]
The thing is that IIS can shutdown your application pool when there are no requests coming in anymore for some given time. IIS doesn't know about Quartz running in the background. 

You can access the setting from IIS Manager application:

* Select application pool hosting your application
* Select Advanced settings from the right
* Change Start Mode from OnDemand to AlwaysRunning

It's a good practice to keep global reference to the scheduler so you can call start on Application_Start and shutdown the scheduler on Application_End

**Comment by keshimp** [comment]
Quick update to add to this thread:

Having it set to "AlwaysRunning" didn't change the outcome - I had tasks fail quite a bit over the weekend.  Also, I have the global static reference to the schedule in app start and that hasn't seemed to resolve the issue either.  I am adding the shutdown on application end now that I have re-read your post this morning and will be trying that, I am also adding a custom app to try to get some data to see if any processes are hanging, etc when the scheduler stops.  It will take a day or two to get all of the data for that.

Is there any built in/optional verbose logging with quartz that I could turn on to get some extra data?  Logging to a file - event viewer, etc?

Thanks again,

**Comment by lahma** [comment]
Configure you favorite logging framework to log to file. Quartz should pick it up automatically using [LibLog ](https://github.com/damianh/LibLog/wiki).

**Comment by keshimp** [comment]
Update - got data overnight with background tasks.  I tried making a script that would show any hanging tasks in IIS whenever the stoppage happened and got nothing.  I am going to setup logging this afternoon with LibLog.  I also have custom logging setup to detect any SQL errors / Quartz errors / any exception that pass through the app should get logged, and have got nothing of substance.

` JobScheduler.scheduler.Shutdown();` - scheduler being my global static var for the scheduler.

Attached are a couple of screenshots with my settings.

![da3642d4](https://user-images.githubusercontent.com/33849905/42832775-e8a610bc-89b7-11e8-9e06-f6a49ba87b0b.PNG)
![3a15dcc5](https://user-images.githubusercontent.com/33849905/42832776-ebbeda18-89b7-11e8-8659-4252f44f6671.PNG)

**Comment by keshimp** [comment]
Set up some in depth logging for this issue - noticed a couple of things.  

First, I noticed that tasks will reschedule themselves after being run (which is what I expect), but they will also reschedule multiple times in one pass(not what I expect), and sometimes they flat out just don't reschedule but continue to run (??), and then ultimately, they just stop running.

The scheduling process on app start works swimmingly - in fact it works in a way that i'd suspect 100%.  I've traced the problem itself down to (i think) the rescheduling process.  Right now - this is how I reschedule tasks:

This will run the task:
`public static void FireJob(BackgroundTaskType jobType)
        {
            if (!ReschedulableJobs.ContainsKey(jobType))
            {
                return;
            }
            var job = ReschedulableJobs[jobType];
            var key = job.Key;
            scheduler.TriggerJob(key).GetAwaiter().GetResult();
        }`

It adds the job to "reschedulableJobs", and then triggers the job

`        public static async Task<DateTime> GetNextFireTimeAsync(BackgroundTaskType jobType)
        {
            if (!ReschedulableJobs.ContainsKey(jobType))
            {
                return default(DateTime);
            }
            var job = ReschedulableJobs[jobType];

            var triggers = await scheduler.GetTriggersOfJob(job.Key);
            var specificTrigger = triggers.First();
            return specificTrigger.GetFireTimeAfter(DateTime.Now).Value.DateTime.ToLocalTime();

        }`

It then gets the next fire time for the job

`public static async Task RescheduleJobAsync(BackgroundTaskType jobType, DateTime scheduledTime)
        {
            if (!ReschedulableJobs.ContainsKey(jobType))
            {
                return;
            }
            var job = ReschedulableJobs[jobType];
            var doesExist = await scheduler.CheckExists(job.Key);
            if (doesExist)
            {
                CancelRescheduledJob(jobType);
            }

            var trigger = TriggerBuilder.Create().StartAt(scheduledTime).Build();
            RescheduledTriggerKeys[jobType] = trigger.Key;

            await scheduler.ScheduleJob(job, trigger);
        }`

It then finally reschedules the task.  This is where I'm not entirely sure everything is going too hot.  Also not sure why it's rescheduling the same task more than once and then suddenly not rescheduling at all.  I will pull more data for this but hoping someone will see something I haven't with a few snippets here.

**Comment by lahma** [comment]
I think your custom logic can have potentially some error, but it's really hard to say without understanding the whole. It might be even as simple as having no locking when accessing some bookkeeping fields like ReschedulableJobs. Unfortunately I can only suggest adding more logging and maybe create a process dump when the stop happens - if it's a real hang. Maybe add heartbeat logging every one minute that processing is OK and not deadlocked etc.

**Comment by keshimp** [comment]
Hi - I am back with some real logging this time -- 

We have been getting this exception, we finally came across it scouring through the logs of a platform we use called Stackify, and found this exception happening during the execution of a background task of ours.  It has happened around 73 times now, and would certainly line up with the issue w e have been having if our collection is getting blocked. 

`11:03:01.868
ERROR The operation was canceled.
System.OperationCanceledException: The operation was canceled.
  at System.Threading.CancellationToken.ThrowOperationCanceledException
  at System.Threading.SemaphoreSlim.WaitUntilCountOrTimeout
  at System.Threading.SemaphoreSlim.Wait
  at System.Threading.SemaphoreSlim.Wait
  at System.Collections.Concurrent.BlockingCollection`1.TryTakeWithNoTimeValidation`

I stumbled upon this thread after searching this exception: https://stackoverflow.com/questions/22967631/how-can-blockingcollectiont-getconsumingenumerable-throw-operationcanceledex

This seems like some underlying IIS stuff.  We changed over to the async logic and scheduling a day before we started getting these exceptions. 

All of the custom logic we have still applies, any red flags?  Thanks again for all of your help and input

**Comment by lahma** [comment]
Maybe make sure that your code has proper locking primites in place to protect from wrong access to collection.  What I understand from this, it probably has something to do with your own logic. I'm closing this as there is little do on Quartz.NET's side. I hope you find the culprit, and if it's Quartz, please ping back and let's fix it.

**Comment by Hasnaa-Ibraheem-I** [comment]
I found this property that you may find helpful as well regarding the issue in the title of "_Quartz scheduler randomly stops firing tasks_", it is: `org.quartz.scheduler.idleWaitTime`
http://www.quartz-scheduler.org/documentation/2.3.1-SNAPSHOT/configuration.html
The default _idelWaitTime_ is **30 minutes** as per this source: http://www.quartz-scheduler.org/documentation/quartz-2.2.2/configuration/ConfigMain.html 
Which means for any job that fires every 5 minutes for example, given that you have a misfire policy of do nothing this job will not fire, its next schedule time will be in the past and accordingly, when the scheduler is back it will not attempt to fire it.

Another property that I see important to boost performance in general (not related to this specific issue) is this one: `org.quartz.scheduler.batchTriggerAcquisitionMaxCount` in case you want to run multiple triggers at once.
